### ISSUE TITLE
If you write text content for a photo post, then add a photo, the text gets deleted

### ISSUE THREAD

**Issue by maxwelljoslyn** [benjifs/sparkles#29]
To reproduce:
1. Sign in to a MicroPub Server that supports photo posts and the media endpoint.
2. Choose to create a photo post.
3. Type some text in the `content` `<textarea>`.
4. Choose the cloud arrow button for picking a photo.
5. Browse for and choose a photo, then upload it via media endpoint.
6. ﻿After successful upload, click "Post Image."
7. The text typed in step 3 will no longer exist.

This may also apply to tags and alt text written before picking a photo -- I haven't tried that yet.

**Comment by benjifs** [comment]
Thanks for reporting this @maxwelljoslyn. Should be fixed in the latest release

**Comment by maxwelljoslyn** [comment]
Thanks Benji!

There’s some other things I would like to see added to sparkles, which you
might want also — for instance, I want the name field to be available on
basically every single type of post. I don’t think it would hurt your usage
if that was possible, yet it would make sparkles more usable for me.

Or, if the stuff I want isn’t your jam, I could always fork. Some of the
things I want are customized post types for my own usage, or making use of
core Indieweb h-entry post types in different ways than you do — I wouldn’t
expect you to support those.

I don’t know JS/TS super well, but I think I’ve been able to suss out the
places  I would want to make edits. So hopefully I can start landing pull
requests.

Let me know if there’s any particular strategy would like me to follow.
Maybe you and I can discuss at homebrew tomorrow? talk to you soon.

**Comment by benjifs** [comment]
@maxwelljoslyn I could see the use case for adding a "name" field in some post types as an editable field but I'm curious if that conflicts with the specification for [post type discovery](https://indieweb.org/post-type-discovery) and how strict those should be.

For example, sparkles allows you to create a "note". Having an optional `name` field here would then make that post an "article". There are other post types like "bookmark" and "like" where `name` would make sense but currently only "bookmark" supports it. And then there's a "photo" post where a `name` could be added and it still counts as a [photo post](https://indieweb.org/photo) but there seems to be some ongoing discussion on where the caption goes and I already pass that caption in as `content`.

One other thing I had thought of but never really implemented was the ability to edit the raw JSON before it gets sent to a micropub endpoint which would allow for sending additional fields or customizing content but it may be too fragile depending on how someone uses it.

Sort of related but I have a feature in the works that also allows you to fetch a sites' metadata to prefill the name and potentially other fields. This is sort of what it looks like: ![Screenshot 2024-08-29 at 14 49 12](https://github.com/user-attachments/assets/9c439ce6-3a89-484c-a8eb-a43a62ef8728)

**Comment by maxwelljoslyn** [comment]
Oh yeah, adding a name to a note would definitely not be good! For me,
names for photos was the most pressing use case (e.g.
https://www.maxwelljoslyn.com/2024/08/23/2).

I'd be in favor of the raw JSON ability as a power user feature, with no
guarantees from you if the user accidentally sends undesired JSON (though
of course you could improve the UX by checking whether it is valid JSON,
and any other checks you can think of -- "what post type does PTD think
this looks like?" maybe.)

I'm not sure I understand what the metadata fetching is for. I could see
that being useful for Micropub updates, but I don't think Sparkles offers
that, right?

**Comment by maxwelljoslyn** [comment]
Oh, now I understand that it could also be useful for stuff like bookmarks: I see that you have a bookmark in the screenshot above.

Presumably you put in the URL and then Sparkles pulled in the title -- but that's weird, why did it pull in the `<title>`, instead of the `p-name`?

**Comment by maxwelljoslyn** [comment]
All in all, I think it would probably be better if I just forked and added my own functionality. That would mean less work for you as I fiddle around and make JavaScript noob errors, or trample on things that you don't want trampled.

When I get this or that feature working, I can always write a PR to share back with you upstream -- or at the very least do a demo, so you can decide whether you might want the feature(s) yourself.

**Comment by benjifs** [comment]
> For me, names for photos was the most pressing use case.

I went ahead and added this in the latest version. I definitely misunderstood some things when building this app so I limited the inputs each post type had. Reading more about it there's really nothing limiting "title" in a photo post for example.

I added other minor things but just wanted to let you know in case you wanted to see how I did it and compare.

If you do end up working on your own fork and there are some features you think would be good to have here, feel free to submit a PR. And no worries on the JS errors or anything. We can work on that together to make sure whatever gets merged looks ok.
